data-model/src/main/java/io/micronaut/data/model/query/builder/sql/SqlSchemaUtils.java

```diff
@@ -183,22 +199,223 @@ public static List<SqlTableDefinition> getSqlTableDefinitions(@NonNull Persisten
+        if (columnMapping.getDbType() == SqlDbType.UUID) {
+            return uuidMatchesVarchar(columnMetadata);
+        } else if (columnMetadata.type() == Types.NUMERIC) {
+            // Custom sql type name for ORACLE
+            String oracleSqlType = "NUMBER";
+            if (columnMetadata.columnSize() > 0) {
+                oracleSqlType += "(" + columnMetadata.columnSize();
+                if (columnMetadata.decimalDigits() > 0) {
+                    oracleSqlType += "," + columnMetadata.decimalDigits();
+                }
+                oracleSqlType += ")";
+            }
+            return columnMapping.getSqlType(Dialect.ORACLE).equalsIgnoreCase(oracleSqlType);
+        } else if (isOracleBinaryDoubleOrFloat(columnMetadata.typeName())) {
+            return isFloatOrRealOrDouble(columnMapping.getDbType().getType());
+        }
+        return false;
+    }
+
+    private static boolean isOracleBinaryDoubleOrFloat(String typeName) {
+        return "BINARY_DOUBLE".equalsIgnoreCase(typeName) || "BINARY_FLOAT".equalsIgnoreCase(typeName);
+    }
+
+    private static boolean matchMySqlColumn(SqlColumnMapping columnMapping, SqlColumnMetadata columnMetadata) {
+        if (columnMapping.getDbType() == SqlDbType.UUID) {
+            return uuidMatchesVarchar(columnMetadata) ||
+                // For MariaDB
+                (columnMetadata.type() == Types.OTHER && columnMetadata.typeName().equalsIgnoreCase("uuid"));
+        }
+        if (columnMapping.getDbType() == SqlDbType.BOOLEAN) {
+            return columnMetadata.type() == Types.BIT;
+        }
+        if (columnMapping.getDbType() == SqlDbType.JSON) {
+            return columnMetadata.type() == Types.LONGVARCHAR;
+        }
+        return false;
+    }
+
+    private static boolean matchH2Column(SqlColumnMapping columnMapping, SqlColumnMetadata columnMetadata) {
+        if (columnMapping.getDbType() == SqlDbType.BINARY) {
+            return columnMetadata.type() == Types.BLOB;
+        }
+        return false;
+    }
+
+    private static boolean matchPostgresColumn(SqlColumnMapping columnMapping, SqlColumnMetadata columnMetadata) {
+        if (columnMapping.getDbType() == SqlDbType.BOOLEAN) {
+            return columnMetadata.type() == Types.BIT;
+        }
+        return false;
+    }
+
+    private static boolean matchSqlServerColumn(SqlColumnMapping columnMapping, SqlColumnMetadata columnMetadata) {
+        if (columnMapping.getDbType() == SqlDbType.BINARY) {
+            return columnMetadata.type() == Types.VARBINARY;
+        }
+        if (columnMapping.getDbType() == SqlDbType.JSON) {
+            return columnMetadata.type() == Types.NVARCHAR;
+        }
+        return false;
+    }
+
+    private static boolean uuidMatchesVarchar(SqlColumnMetadata columnMetadata) {
+        return columnMetadata.type() == Types.VARCHAR && columnMetadata.columnSize() == 36;
```

Comment:
We have one potential issue for Oracle and MySql dialects where we store UUID as varchar(36) and convert to string when binding. But, Oracle can represent UUID as `RAW(16)` and MySql as `BINARY(16)` so we cannot support that and column types won't match. This is an example of issues we could have and didn't want to change binding/saving code not wanting to break current behavior.